chore: refactor MapperGenerator to cache outputs.

TimothyMakkison opened the pull request:
# Refactored `MapperGenerator` to use pipeline caching

## Description
Refactored `MapperGenerator` to use pipeline caching based upon technique used [here](https://github.com/CommunityToolkit/dotnet/blob/main/src/CommunityToolkit.Mvvm.SourceGenerators/Input/RelayCommandGenerator.cs#L23). 

Because of the use of `Compilation` it is hard to cache any of the `Descriptor` logic, instead I moved the bulk of the generation and creation logic into a single pipeline step. All descriptors are calculated together yielding `(ImmutableArray<(MemberDeclarationSyntax, string)>` and `ImmutableArray<Diagnostic>` warnings/errors. The warning are reported and the mapping name pairs are split into individual values. Finally `RegisterImplementationSourceOutput` runs `NormalizeWhitespace`, creates a `CompilationUnitSyntax`, gets `SourceText.From` and adds the source. 

Hopefully subsequent compilations will recognise that `(MemberDeclarationSyntax, string)` is the same, skipping `AddSource` and the costly `NormalizeWhitespace` (usually 20% of the memory usage).

This doesn't appear to change any of the benchmarks but should improve IDE responsiveness 🤞 . 

Please double check my logic. Is there a way to see if caching is occuring? Source generators are still a black box to me. I might have to return `EquatableArrays` from `BuildDescriptors`

Helps #72

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

latonz commented on src/Riok.Mapperly/MapperGenerator.cs:
```diff
@@ -22,39 +22,67 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 
         var compilationAndMappers = context.CompilationProvider.Combine(mapperClassDeclarations.Collect());
-        context.RegisterImplementationSourceOutput(compilationAndMappers, static (spc, source) => Execute(source.Left, source.Right, spc));
+        IncrementalValueProvider<(
+            ImmutableArray<(MemberDeclarationSyntax, string)> Mappers,
+            ImmutableArray<Diagnostic> Diagnostics
+        )> mappersWithErrors = compilationAndMappers.Select((x, token) => BuildDescriptors(x.Left, x.Right, token));
```
I think it would make sense to create a class for the Mappers / Diagnostic combination.

latonz commented on src/Riok.Mapperly/Emit/SourceEmitter.cs:
```diff
@@ -8,16 +8,19 @@ namespace Riok.Mapperly.Emit;
         member = WrapInClassesAsNeeded(descriptor.Symbol, member);
-        member = WrapInNamespaceIfNeeded(descriptor.Namespace, member);
+        return WrapInNamespaceIfNeeded(descriptor.Namespace, member);
+    }
```
Couldn't the `CompilationUnitSyntax` be returned here, just without the `NormalizeWhiteSpace`? Then the MapperGenerator could call `NormalizeWhiteSpace` in the very last step.

latonz reviewed (changes requested):
Thank you for opening this PR.

To test each intermediate step you can use `Microsoft.CodeAnalysis.GeneratorRunResult.TrackedSteps` / `Microsoft.CodeAnalysis.GeneratorRunResult.TrackedOutputSteps` (eg. `Riok.Mapperly.Tests.TestHelper.Generate().GetRunResult()..Results[0].TrackedSteps`). We should definitely test whether the caching between the steps works. Eg. the reason on each step can be used to assert this.

TimothyMakkison commented:
Thanks for the information, I wish incremental generators had better documentation. 

- Added incremental source generator tests
- Added `MapperResult` and `MapperNode` types, I've opted to use `IsEquivalentTo` for the node comparisons, I'm not sure if this is a good idea, it's probably slow in large projects/mappers.

latonz commented on test/Riok.Mapperly.Tests/Generator/IncrementalGeneratorTest.cs:
```diff
@@ -0,0 +1,161 @@
+    [Fact]
+    public void AppendingUnrelatedTypeDoesNotRegenerateOriginal()
+    {
+        var source = TestSourceBuilder.CSharp(
+            """
+            using Riok.Mapperly.Abstractions;
+
+            namespace Test.A
+            {
+                [Mapper]
+                internal partial class FooBarMapper
+                {
+                    internal partial string FooToBar(string value);
+                }
+            }
+            """
+        );
```
Instead of these mapper sources couldn't `Riok.Mapperly.Tests.TestSourceBuilder.Mapping("string", "string")` be used?

latonz commented on test/Riok.Mapperly.Tests/Generator/IncrementalGeneratorTest.cs:
```diff
@@ -0,0 +1,161 @@
+
+        var driver1 = TestHelper.TrackingGeneratorDriver().RunGenerators(compilation1);
+
+        var compilation2 = compilation1.AddSyntaxTrees(CSharpSyntaxTree.ParseText("struct MyValue {}", CSharpParseOptions.Default));
```
I think it would make sense to add a `SyntaxTree` method (and probably additional overloads to `Mapping`: `MappingSyntaxTree`) to the `TestSourceBuilder` to improve the readability of the tests. Eg. `TestSourceBuilder.SyntaxTree("struct MyValue {}")`. This would also improve IDE support as some IDE's recognise the `StringSyntax` `csharp` (eg. Rider), which `ParseText` doesn't have.

latonz commented on test/Riok.Mapperly.Tests/TestHelper.cs:
```diff
@@ -80,4 +71,27 @@ private static CSharpCompilation BuildCompilation(NullableContextOptions nullabl
         return CSharpCompilation.Create("Tests", syntaxTrees, references, compilationOptions);
     }
+
+    public static GeneratorDriver TrackingGeneratorDriver()
```
Instead of this method, I think `GenerateTracked(Compilation)` could be exposed, which calls `RunGenerators` on the driver (each caller of this method calls `RunGenerators` anyway).

latonz commented on test/Riok.Mapperly.Tests/Generator/IncrementalGeneratorTest.cs:
```diff
@@ -0,0 +1,161 @@
+        AssertRunResults(driver2, IncrementalStepRunReason.New);
+    }
+
+    private static void AssertRunResults(GeneratorDriver driver, params IncrementalStepRunReason[] runReasons)
```
Diagnostics should probably also be tested. FYI the `IncrementalValuesProvider` has a `WithTrackingName` method to set step tracking names.

TimothyMakkison replied:
`WithTrackingName` is not available in older versions of roslyn. I could use the auto generated name or wrap everything in `#if` preproccessors

TimothyMakkison replied:
No idea how to use `WithTrackingName` with older roslyn versions

latonz replied:
You could probably just add an extension method `IncrementalValueProvider<TSource> WithTrackingName<TSource>(this IncrementalValueProvider<TSource> source, string name) => source;` to mock this for older roslyn versions.

TimothyMakkison replied:
My main issue is getting the tests to pass. I can use preprocessors/mock `WithTrackingName`, but I can't get the tests to pass.

For instance if I wrap the tests in Net_7_And_Above the tests will still run for a version of Roslyn that does not add the tracking name

latonz replied:
Hmmm, but the unit tests only run for the latest Roslyn version supported by Mapperly, don't they?

TimothyMakkison replied:
That's what I thought too 😆 

I'll double check I probably made a nistake

latonz commented on src/Riok.Mapperly/MapperGenerator.cs:
```diff
@@ -22,39 +22,63 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 
         var compilationAndMappers = context.CompilationProvider.Combine(mapperClassDeclarations.Collect());
-        context.RegisterImplementationSourceOutput(compilationAndMappers, static (spc, source) => Execute(source.Left, source.Right, spc));
+        var mappersWithErrors = compilationAndMappers.Select((x, token) => BuildDescriptors(x.Left, x.Right, token));
```
I'd rename to `mappersWithDiagnostics` as these are not necessarily errors.

latonz replied:
This select many lambda could also be static, couldn't it?

latonz replied:
Would rename `token` to `ct` or `cancellationToken` as only token seems more syntax related to me in this context.

latonz commented on src/Riok.Mapperly/MapperGenerator.cs:
```diff
@@ -22,39 +22,63 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+        context.ReportDiagnostics(mappersWithErrors.Select(static (source, _) => source.Diagnostics));
+
+        // split into mapper name pairs
+        var mappers = mappersWithErrors.SelectMany((x, _) => x.Mappers);
```
This select many lambda could also be static, couldn't it?

latonz commented on src/Riok.Mapperly/MapperGenerator.cs:
```diff
@@ -22,39 +22,63 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+            mappers,
+            static (spc, source) =>
+            {
+                spc.AddSource(source.FileName, SourceText.From(source.Body.NormalizeWhitespace().ToFullString(), Encoding.UTF8));
```
A lot going on on this line, probably split it into multiple lines to increase readability.

TimothyMakkison commented on src/Riok.Mapperly/MapperResults.cs:
```diff
@@ -0,0 +1,33 @@
+
+    public bool Equals(MapperResults other)
+    {
+        return Mappers.SequenceEqual(other.Mappers) && Diagnostics.SequenceEqual(other.Diagnostics);
```
`MapperResult` is a value type therfore `ReferenceEquals` doesn't work. i can make it a class

latonz replied:
My bad, thank you the explanation. IMO struct is a good fit here.

latonz commented on src/Riok.Mapperly/Descriptors/DescriptorBuilder.cs:
```diff
@@ -35,13 +35,13 @@ WellKnownTypes wellKnownTypes
     }
 
-    public MapperDescriptor Build()
+    public (MapperDescriptor descriptor, List<Diagnostic> errors) Build()
```
IMO the name errors is not appropriate here, as it can contain any diagnostics.

latonz commented on src/Riok.Mapperly/MapperGenerator.cs:
```diff
@@ -22,39 +22,66 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+        );
+
+        // output the diagnostics
+        context.ReportDiagnostics(mappersWithDiagnostics.Select(static (source, _) => source.Diagnostics).WithTrackingName("diagnostics"));
```
Since `WithTrackingName` is not available in older Roslyn versions we could probably just add an extension method `IncrementalValueProvider<TSource> WithTrackingName<TSource>(this IncrementalValueProvider<TSource> source, string name) => this;` to mock this for older roslyn versions.

latonz commented on src/Riok.Mapperly/MapperNode.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        unchecked
+        {
+            return (Body.GetHashCode() * 397) ^ FileName.GetHashCode();
```
Couldn't this return another hash code for entities which are actually equal? (If the `Body` is another instance which is equal to each other?)

TimothyMakkison replied:
Good question, it might.

I don't think incremental generators use the objects hash code or `SyntaxNode` has some custom `GetHashCode` method. Otherwise the tests would never pass and it wouldn't be possible to cache `SyntaxNode`, defeating the purpose of the incremental source generator.

latonz replied:
IMO not using the hash code is an implementation detail of how Roslyn calls the source generators which Mapperly should not rely on. IMO the fix here is also pretty simple:
* Instead of using the `Body.GetHashCode()` use a simpler hash code, eg. `Body.Members.Count.GetHashCode()`
* Remove the hash code part of the body completely

latonz commented on src/Riok.Mapperly/MapperResults.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public static readonly MapperResults Empty = new(ImmutableArray<MapperNode>.Empty, ImmutableArray<Diagnostic>.Empty);
+
+    public MapperResults(ImmutableArray<MapperNode> mappers, ImmutableArray<Diagnostic> diagnostics)
```
Couldn't we simply use `IReadOnlyCollection` instead?

latonz commented on src/Riok.Mapperly/MapperResults.cs:
```diff
@@ -0,0 +1,50 @@
+            return hash;
+        }
+
+        static int Combine(int h1, int h2)
```
We could probably backport the [HashCode](https://learn.microsoft.com/en-us/dotnet/api/system.hashcode) struct?

latonz commented on test/Riok.Mapperly.Tests/Generator/IncrementalGeneratorTest.cs:
```diff
@@ -0,0 +1,121 @@
+[UsesVerify]
+public class IncrementalGeneratorTest
+{
+    private const string AddMappersStep = "ImplementationSourceOutput";
+    private const string ReportDiagnosticsStep = "diagnostics";
```
I think it could be helpful to define these names as consts inside `Riok.Mapperly` and reference them from the MapperGenerator and here from the tests.

latonz commented on test/Riok.Mapperly.Tests/Generator/IncrementalGeneratorTest.cs:
```diff
@@ -0,0 +1,121 @@
+
+        var secondary = TestSourceBuilder.SyntaxTree(
+            """
+using Riok.Mapperly.Abstractions;
+
+namespace Test.B
+{
+    [Mapper]
+    internal partial class BarFooMapper
+    {
+        internal partial string BarToFoo(string value);
+    }
+}
+"""
```
Indent as other code...

TimothyMakkison commented:
Added `ImmutableEquatableArray` from `System.Text.Json`, its not needed but its more idiomatic IMO, happy to remove it.
